Implement custom exceptions

Pipas1903 opened the pull request:
I did a lot more, however i don't remember anymore 🤷🏽‍♀️ I am being very descriptive here. Also here's the ticket link: www.ticket-link.com .
I'm explaining what I developed here.

Pipas1903 committed:
added some missing elements and anotations

Pipas1903 committed:
changes

Pipas1903 committed:
Merge branch 'master' into filipa

Pipas1903 committed:
small improvements

GenesisMRL reviewed (commented):
it's fish!

rafaeltmiranda commented on docker-compose.yml:
```diff
@@ -0,0 +1,25 @@
+    volumes:
+      - local-data:/rabbitmq
+    networks:
+      - basic_net
```
Having a blank line at the end of each file is a standard that GitHub will impose. From now on try to never forget that ;)

rafaeltmiranda commented on logs/archived/spring-boot-logger-2022-03-21.0.log:
```diff
@@ -0,0 +1,533 @@
+2022-03-21 12:52:42,472 INFO org.springframework.boot.StartupInfoLogger [main] Starting SpaceshipRentApplication using Java 17.0.1 on Filipas-MacBook-Pro.local with PID 49209 (/Users/pipinha/Back End/spaceship-rent-temp/target/classes started by pipinha in /Users/pipinha/Back End/spaceship-rent-temp)
```
please add the logs folder to your `.gitignore` file

rafaeltmiranda commented on logs/archived/spring-boot-logger-2022-03-22.0.log:
```diff
@@ -0,0 +1,160 @@
+2022-03-22 16:04:00,506 INFO org.hibernate.validator.internal.util.Version [background-preinit] HV000001: Hibernate Validator 6.2.2.Final
```
Same here - `.gitignore`

rafaeltmiranda commented on logs/archived/spring-boot-logger-2022-03-23.0.log:
```diff
@@ -0,0 +1,141 @@
+2022-03-23 11:15:35,039 INFO org.hibernate.validator.internal.util.Version [background-preinit] HV000001: Hibernate Validator 6.2.2.Final
```
Same here - `.gitignore`

rafaeltmiranda commented on logs/archived/spring-boot-logger-2022-03-24.0.log:
```diff
@@ -0,0 +1,2097 @@
+2022-03-24 11:57:52,584 INFO org.hibernate.validator.internal.util.Version [background-preinit] HV000001: Hibernate Validator 6.2.2.Final
```
Same here - `.gitignore`

rafaeltmiranda commented on logs/archived/spring-boot-logger-2022-03-25.0.log:
```diff
@@ -0,0 +1,1649 @@
+2022-03-25 10:59:53,658 INFO org.hibernate.validator.internal.util.Version [background-preinit] HV000001: Hibernate Validator 6.2.2.Final
```
Same here - `.gitignore`

rafaeltmiranda commented on pom.xml:
```diff
@@ -68,6 +68,30 @@
 <!--            <version>8.4.0</version>-->
 <!--        </dependency>-->
```
We need to uncomment this dependency and solve the test issues, to have flyway working as expected

rafaeltmiranda commented on src/main/java/com/mindera/school/spaceshiprent/components/EmailConsumer.java:
```diff
@@ -0,0 +1,58 @@
+                                        "Hope you enjoy our spaceships and customer service :)")
+                                .put(Emailv31.Message.CUSTOMID, "spaceshipRentMailingSystem")));
+        response = client.post(request);
+        System.out.println(response.getStatus());
```
We don't sout stuff. Ether log it ou remove it ;)

rafaeltmiranda commented on src/main/java/com/mindera/school/spaceshiprent/components/EmailConsumer.java:
```diff
@@ -0,0 +1,58 @@
+public class EmailConsumer {
+
+    private final Logger LOGGER = LoggerFactory.getLogger(EmailConsumer.class);
+    private final String KEY = "26d4ec8cfda8f82be1c1a67d8f3e99a4";
```
This should be under properties on the application.yml (or in the specific profile properties if applicable) and you should avoid having the explicit Key and secret in the codebase. Use environment variables instead.

rafaeltmiranda commented on src/main/java/com/mindera/school/spaceshiprent/components/EmailConsumer.java:
```diff
@@ -0,0 +1,58 @@
+
+    private final Logger LOGGER = LoggerFactory.getLogger(EmailConsumer.class);
+    private final String KEY = "26d4ec8cfda8f82be1c1a67d8f3e99a4";
+    private final String SECRET_KEY = "39abedbf3999b62c2f072a70f3f0fdb8";
```
Same

rafaeltmiranda commented on src/main/java/com/mindera/school/spaceshiprent/components/EmailConsumer.java:
```diff
@@ -0,0 +1,58 @@
+        MailjetClient client;
+        MailjetRequest request;
+        MailjetResponse response;
+        client = new MailjetClient(KEY, SECRET_KEY, new ClientOptions("v3.1"));
```
You should abstract Third-party specific implementation from here. You could have a `EmailGateway` interface, and a `MailjetGateway` which will implement the `EmailGateway` and it should the only class that knows how Mailjet works. Any other should work regardless of the third-party implementation.

rafaeltmiranda commented on src/main/java/com/mindera/school/spaceshiprent/exception/exceptionHandlers/ValidationExceptionHandler.java:
```diff
@@ -0,0 +1,30 @@
+package com.mindera.school.spaceshiprent.exception.exceptionHandlers;
```
Delete this :)

rafaeltmiranda replied:
And having only one handler (as expected) you don't need a package specially for that

rafaeltmiranda commented on src/main/java/com/mindera/school/spaceshiprent/exception/model/ValidationError.java:
```diff
@@ -0,0 +1,17 @@
+package com.mindera.school.spaceshiprent.exception.model;
```
Move this to the `SpaceshipRentError` and delete this file

rafaeltmiranda commented on src/main/java/com/mindera/school/spaceshiprent/persistence/entity/UserEntity.java:
```diff
@@ -1,26 +1,11 @@
-import lombok.NoArgsConstructor;
-import lombok.Setter;
-import lombok.ToString;
+import lombok.*;
```
we don't want `*` imports. it's like having to talk to a friend and all of a sudden all your friends are in to talk with you :D

rafaeltmiranda commented on src/main/java/com/mindera/school/spaceshiprent/service/RabbitBroker.java:
```diff
@@ -0,0 +1,9 @@
+@Service
+@RequiredArgsConstructor
+public class RabbitBroker {
+}
```
If this class is not needed, please delete it. If it is, complete the implementation..
AND....
extra line please :D

rafaeltmiranda commented on src/main/java/com/mindera/school/spaceshiprent/service/spaceship/SpaceShipServiceImpl.java:
```diff
@@ -38,25 +40,23 @@ public SpaceShipDetailsDto createSpaceShip(CreateOrUpdateSpaceshipDto createOrUp
-                .orElse(null);
+        return converter.convertToSpaceShipDetailsDto(spaceShipRepository.findById(id)
+                .orElseThrow(() -> new SpaceshipNotFoundException(String.format(ErrorMessageConstants.SPACESHIP_NOT_FOUND, id)) {
+                })
```
what's the purpose of this `{ }` after instantiating the exception? not following :D

rafaeltmiranda commented on src/main/java/com/mindera/school/spaceshiprent/service/spaceship/SpaceShipServiceImpl.java:
```diff
@@ -38,25 +40,23 @@ public SpaceShipDetailsDto createSpaceShip(CreateOrUpdateSpaceshipDto createOrUp
         Optional<SpaceshipEntity> spaceShipEntityOptional = spaceShipRepository.findById(id);
 
-        if(spaceShipEntityOptional.isPresent()){
+        if (spaceShipEntityOptional.isPresent()) {
```
I would rather validate the exception first and throw the exception right away. Your if will have less noise and most of the logic will be out of any if (simpler to read)

rafaeltmiranda commented on src/main/java/com/mindera/school/spaceshiprent/service/user/UserServiceImpl.java:
```diff
@@ -20,10 +20,13 @@
 
     private final UserConverter converter;
     private final UserRepository userRepository;
+    //private final EmailSender emailSender;
```
Ether complete the implementation or remove it for now :)

rafaeltmiranda commented on src/main/java/com/mindera/school/spaceshiprent/service/user/UserServiceImpl.java:
```diff
@@ -20,10 +20,13 @@
     public UserDetailsDto createUser(CreateOrUpdateUserDto createOrUpdateUserDto) {
         UserEntity userEntity = converter.convertToEntity(createOrUpdateUserDto);
+        String emailingInfo = userEntity.getEmail() + " " + userEntity.getName();
+        // emailSender.send(emailingInfo);
```
Same

rafaeltmiranda commented on src/main/java/com/mindera/school/spaceshiprent/service/user/UserServiceImpl.java:
```diff
@@ -51,8 +54,12 @@ public UserDetailsDto updateUserById(Long id, CreateOrUpdateUserDto createOrUpda
+        throw new UserNotFoundException(String.format(ErrorMessageConstants.USER_NOT_FOUND, id));
+    }
+
+    public void deleteUserById(Long id) {
```
Since it will stay here, let's do it right, with validations exceptions and logs ;)

rafaeltmiranda commented on src/main/java/com/mindera/school/spaceshiprent/util/LoggerMessages.java:
```diff
@@ -0,0 +1,18 @@
+package com.mindera.school.spaceshiprent.util;
```
👌

rafaeltmiranda commented on src/main/resources/application-dev.yml:
```diff
@@ -0,0 +1,5 @@
+  show-sql: true
+  properties:
+    hibernate:
+      format_sql: true
```
Extra line

rafaeltmiranda commented on src/main/resources/application.yml:
```diff
@@ -7,19 +7,21 @@ spring:
+    host: localhost
+    port: 5672
+    username: guest
+    password: guest
```
extra line :D

rafaeltmiranda commented on src/main/resources/logback.xml:
```diff
@@ -0,0 +1,41 @@
+        <appender-ref ref="Console" />
+    </root>
+
+</configuration>
```
extra line :D

rafaeltmiranda reviewed (changes requested):
I've left a lot of comments for you (the whole class @mindera-school/class-2021-2022 )

Next steps to merge this PR:
* Fix all the things I pointed
* Add tests in order to raise the coverage by over 80%
* Be happy 😆 

That's it for now! Good luck! Love you all 💛

Pipas1903 committed:
minor corrections

rmzaoo committed:
added postgres to docker-compose (#6)

Co-authored-by: Filipa Silva <92788310+Pipas1903@users.noreply.github.com>

rafaeltmiranda commented:
> I think there might be some tests failing with flyway, right? Maybe you need to replicate flyway properties to the tests' application.yml. Did you run the tests?

See that the `Build` check failed. You can see there what's the reason. 

@mindera-school/class-2021-2022 Please **always** run `mvn clean install` on your computer before pushing into the remote.

rafaeltmiranda commented on src/main/java/com/mindera/school/spaceshiprent/emailing/EmailConsumer.java:
```diff
@@ -1,4 +1,4 @@
-package com.mindera.school.spaceshiprent.components;
+package com.mindera.school.spaceshiprent.emailing;
```
I really don't like the package name. I'd rather prefer the old one, at least as the main package for components.

Pipas1903 committed:
created some user converter tests

Pipas1903 committed:
flyway tests fix

Pipas1903 committed:
added logs to gitignore

joao-moutinho committed:
spaceship controller test and spaceship service test (#7)

GenesisMRL committed:
added a couple of tests in rentController (#8)

rmzaoo committed:
flyway application.yml fix (#10)

rafaeltmiranda committed:
exclude some packages from being considered for coverage calculation (#11)

Add necessary configs to exclude the following packages from being considered by coverage calcs:

dto/*
persistence/entity/*
config/*
exception/*
test/*

Overall coverage has raised to 35,5%

rafaeltmiranda commented:
Added a commit with some coverage exclusions from classes that don't make sense to be considered for coverage. You have no excuses now 😉

GonVieira committed:
Added Spaceship Converter tests. (#9)

Co-authored-by: rmzaoo <mazer.yt.email@gmail.com>

rafaeltmiranda commented:
`38.7%Estimated after merge`
It's improving peeps 💪 
Keep going... go go go 🏃

RuicTeixeira committed:
Changes on updateRent method, addition of setters missing, changes on Exceptions in createRent (#13)

* fixed bugs

* bug fix

* added test and fix in implementation

* minor fixes

* spaces fix

Pipas1903 committed:
fixed error related to environment variable

Pipas1903 committed:
changed mail sender keys to environment variables

rmzaoo committed:
More Unit Tests  (#19)

* finished unit Servic Rent Tests

* spaceship service unit tests finished

Pipas1903 committed:
Revert "Changes on updateRent method, addition of setters missing, changes on Exceptions in createRent (#13)"

This reverts commit 18d76f0c110c7c4cc8985ebe208b3cb54f401de7.

Pipas1903 committed:
revert jwt and bcrypt
